### PR TITLE
Toggle performance mode

### PR DESCRIPTION
**Pull request by sohamazing**
toggle performance mode from GUI

### CODE REVIEW

**Comment by ianohara** [comment]
Hey @sohamazing , I'm taking a look at this but before I get too deep: is there info on what specific performance issues we're trying to fix?

EG: If we run a multi hour acquisition, we run out of memory.  Or gui updates grind to a halt.  Or in some other setup other specific problems present themselves.  The more information the better!

The reason I ask is that I'm wondering if we can fix the root problem, not sacrifice gui utility, and avoid the sorta-hard-to-maintain special case of a toggle for performance mode.  At first glance it seems like our issue is sending all the frames to the various nepari viewers during an acquisition.  If the problem is that we run out of memory as we get a lot of frames, or the gui grinds to a halt with a lot of frames, maybe we can just keep the N last frames instead?  Or for the gui viewers, use hugely down sized images?  I dunno - it depends on what the actual performance issues we're seeing are.  So any if you have on that would be really helpful before reviewing this!

File: software/control/gui_hcs.py [hunk 1]
```diff
@@ -425,24 +427,25 @@ def setupImageDisplayTabs(self):
...
-            self.imageDisplayTabs.addTab(self.napariMultiChannelWidget, "Multichannel Acquisition")
-        else:
-            self.imageArrayDisplayWindow = core.ImageArrayDisplayWindow()
-            self.imageDisplayTabs.addTab(self.imageArrayDisplayWindow.widget, "Multichannel Acquisition")
-
-        if SHOW_TILED_PREVIEW:
-            if USE_NAPARI_FOR_TILED_DISPLAY:
-                self.napariTiledDisplayWidget = widgets.NapariTiledDisplayWidget(self.objectiveStore, self.contrastManager)
-                self.imageDisplayTabs.addTab(self.napariTiledDisplayWidget, "Tiled Preview")
+        if not self.performance_mode:
```
[the remark this answers is not in the document]
Comment (ianohara) [reply software/control/gui_hcs.py:430]: I think the solution may be to always initialize the napari widgets (if the corresponding NAPARI config dictate it), but then just connect/disconnect them for performance mode in the paths from`__init__` like you do with the toggle.

File: software/control/core.py [hunk 2]
```diff
@@ -2320,19 +2322,20 @@ def save_image(self, image, file_ID, config, current_path):
...
-        if USE_NAPARI_FOR_MULTIPOINT or USE_NAPARI_FOR_TILED_DISPLAY:
-            if not self.init_napari_layers:
-                print("init napari layers")
-                self.init_napari_layers = True
-                self.napari_layers_init.emit(image.shape[0],image.shape[1], image.dtype)
-            self.napari_layers_update.emit(image, i, j, k, config_name)
-        if USE_NAPARI_FOR_MOSAIC_DISPLAY and k == 0:
-            print(f"Updating mosaic layers: x={self.navigationController.x_pos_mm:.6f}, y={self.navigationController.y_pos_mm:.6f}")
-            self.napari_mosaic_update.emit(image, self.navigationController.x_pos_mm, self.navigationController.y_pos_mm, k, config_name)
+        if not self.performance_mode:
```
Comment (ianohara) [inline software/control/core.py:2325]: This may be something that's because of my relative unfamiliarity with qt, but since we're disconnecting the napari widgets when in performance mode, can we just leave the emits here and not need this `if not self.performance_mode` check?  Since there won't be any slots listening it's hopefully effectively no-op performance-wise.

Comment (ianohara) [reply software/control/core.py:2325]: If not, is this sufficient for performance mode and we can just not need to deal with the signal/slot connect/disconnect dance?  AKA just don't emit when in performance mode?

File: software/control/gui_hcs.py [hunk 3]
```diff
@@ -110,6 +111,7 @@ def __init__(self, is_simulation=False, performance_mode=False, *args, **kwargs)
             self.waitForMicrocontroller()
             if ENABLE_SCAN_GRID:
                 self.multiPointWidgetGrid.init_z()
+            self.multiPointWidget2.init_z()
```
Comment (ianohara) [inline software/control/gui_hcs.py:114]: Since this isn't related to the performance mode change, it'd be nice to put it in its own pull request.  I think we're aiming for development speed, though, so probably fine for this one!  But in the future having it be as a separate bug fix pull request would be nice.

File: software/control/gui_hcs.py [hunk 4]
```diff
@@ -679,6 +689,10 @@ def makeConnections(self):
 
                 self.multipointController.napari_layers_init.connect(self.napariMultiChannelWidget.initLayers)
                 self.multipointController.napari_layers_update.connect(self.napariMultiChannelWidget.updateLayers)
+                self.napari_connections['napariMultiChannelWidget'] = [
```
Comment (ianohara) [inline software/control/gui_hcs.py:692]: Instead of keeping track of the napari_connections, It might be a bit easier, and more generalizable to other performance fixes to instead just make all the performance-mode-only connections and disconnections in a single method.  As one possible way of doing this, we could get rid of `self.nepari_connections` and have something like your `setup*Connections` method, and move all the `connect`/`disconnect` there.  For instance:

```
def updatePerformanceWidgetConnections():
    if self.performance_mode:
        # Disconnect all the performance mode widgets from the multipointController
        self.multipointController.napari_layers_init.disconnect(self.napariMultiChannelWidget.initLayers)
        self.multipointController.napari_layers_update.disconnect(self.napariMultiChannelWidget.updateLayers)
        # ... the same for all the other performance mode widgets
    else:
       # Not in performance mode, setup all the connections to the above widgets
        self.multipointController.napari_layers_init.connect(self.napariMultiChannelWidget.initLayers)
        self.multipointController.napari_layers_update.connect(self.napariMultiChannelWidget.updateLayers)
       # ... all the others ...
```

Then we just call self.updatePerformanceModeConnections() in `__init__` and when we get a performance mode toggle update. 

There might be much nicer ways to do this in qt as well.  For instance, if you're in c++ it looks like `QMetaObject::disconnect` might be nice - it disconnects all connections between two `QObject`.  There also might be a way to block signals from one QObject to another.  I'm not too qt familiar, though, so would need to investigate more!

**Comment by ianohara** [comment]
The tl;dr of the review:
* Move all the performance mode specific connect/disconnect into a helper method.
* Call that at the end of widget setup in `__init__`, and when we toggle performance.  This lets us get rid of the `self.napari_connections` dict because all the connections/disconnections for performance mode happen in once place, and lets us expand performance mode in the future easily.
* As long as it is essentially a qt no-op to emit on an unconnected signal, remove the performance check and just emit (and let the connect/disconnect handle making it a noop)
* Remove unused code
* Add `self.attribNames = None` in `__init__` so we can always check for `if self.attribName` and avoid using `hasattr` or any of the `*attr*` methods with strings.

**Review by ianohara — CHANGES_REQUESTED** [review]
The tl;dr of the review:
* Move all the performance mode specific connect/disconnect into a helper method.
* Call that at the end of widget setup in `__init__`, and when we toggle performance.  This lets us get rid of the `self.napari_connections` dict because all the connections/disconnections for performance mode happen in once place, and lets us expand performance mode in the future easily.
* As long as it is essentially a qt no-op to emit on an unconnected signal, remove the performance check and just emit (and let the connect/disconnect handle making it a noop)
* Remove unused code
* Add `self.attribNames = None` in `__init__` so we can always check for `if self.attribName` and avoid using `hasattr` or any of the `*attr*` methods with strings.


Also, comments explaining the "why" of things might be helpful!  EG for performance mode.

File: software/control/gui_hcs.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sohamazing) [reply software/control/gui_hcs.py:430]: there is a 'performance mode' we can load directly when we start the program, either with --performance-mode in the command_line or PERFORMANCE_MODE = True in the config files. 
However, this performance mode is meant to be similar but we can turn it on and off from the GUI inbetween acqusitions.

[the remark this answers is not in the document]
Comment (sohamazing) [reply software/control/gui_hcs.py:430]: this is correct, except hongquan mentioned that there will never be a time when we have both performance modes on at the same time. I think the easiest fix would be to remove the toggle performance mode button if we have already launched the GUI in perfromance mode from the config

[the remark this answers is not in the document]
Comment (sohamazing) [reply software/control/gui_hcs.py:430]: The only goal for the performance mode was for speed improvements caused by the napari widgets.

**Comment by ianohara** [comment]
We talked offline.  Here are my notes from the conversation:
* Rename the config/argument based mode to LIVE_VIEW_ONLY or something would help, and in that case just remove the button for performance mode as Soham mentioned above
* Instead of my suggested updatePerformanceWidgetConnections which would require a lot of re-checking config, add a helper called def performance_connection(self, slog, signal, connected): that maintains a self._performance_connections list of (slot, signal) and can be used to connect/disconnect them (or something along these lines).  This is essentially what Soham has but with wrappers around manipulating napari_connections so it isn't all inline like in the pr now.
* Most the other changes seem reasonable to do and Soham will make 'em

File: software/control/core.py (hunk 2 above)
Comment (sohamazing) [reply software/control/core.py:2373]: youre correct, this is redundant

File: software/control/gui_hcs.py [hunk 5]
```diff
@@ -528,6 +532,14 @@ def setupLayout(self):
         layout.addWidget(self.sampleSettingsWidget)
         layout.addWidget(self.navigationViewer)
 
+        # Add performance mode toggle button
```
Comment (ianohara) [inline software/control/gui_hcs.py:535]: the code sorta tells us this already.  For comments, a good rule is that you should use them to explain non-obvious things and/or provide critical context.

For instance, here you could have:
```
# live_only_mode is a config mode that sets up the gui to only show the live view.  In that case, the performance
# mode toggle doesn't do anything since its purpose is to toggle on/off the non-live views.  So we hide the
# performance mode toggle when the system boots in live_only_mode
```

File: software/control/gui_hcs.py [hunk 6]
```diff
@@ -742,6 +706,160 @@ def makeConnections(self):
 
         self.camera.set_callback(self.streamHandler.on_new_frame)
 
+    def makeNapariConnections(self):
```
Comment (ianohara) [inline software/control/gui_hcs.py:709]: This is better than having them spread throughout, but still a bit confusing since it initializes non-Napari stuff when we aren't using napari for the live view.

We can go with it for now.

File: software/control/gui_hcs.py [hunk 7]
```diff
@@ -742,6 +706,160 @@ def makeConnections(self):
...
+                                # Connection might not exist, which is fine
+                                pass
+                        else:
+                            try:
+                                signal.connect(slot)
+                            except TypeError:
+                                # Connection might already exist, which is fine
+                                pass
+
+    def toggleNapariTabs(self):
```
Comment (ianohara) [inline software/control/gui_hcs.py:842]: toggle normally implies switching to the opposite state.  Something like `setNapariTabsForPerformanceMode(self)` or soemthing might be better.

File: software/control/widgets.py [hunk 8]
```diff
@@ -3675,12 +3694,15 @@ def init_z(self, z_pos_mm=None):
 
         self.entry_minZ.blockSignals(True)
         self.entry_maxZ.blockSignals(True)
-        self.navigationController.zPos.disconnect(self.update_z_min)
-        self.navigationController.zPos.disconnect(self.update_z_max)
+        try:
+            self.navigationController.zPos.disconnect(self.update_z_min)
+            self.navigationController.zPos.disconnect(self.update_z_max)
+        except TypeError:
```
Comment (ianohara) [inline software/control/widgets.py:3700]: Do we really just wanna ignore this?

**Review by ianohara — APPROVED** [review]
I think we can just merge this and then make follow up changes as needed.  It works as expected, and it's mostly following existing code conventions.

**Comment by ianohara** [comment]
The one request is that it'd be nice to have a better explanation as the merge commit - so maybe change the title + description to be more clear about what this does?
